app/notifiers/notifier/base.rb

```diff
@@ -0,0 +1,72 @@
+# The base class of notifiers. This is meant to be called by the Notifications Framework
+#
+# @api notifications
+class Notifier::Base
+  include ApplicationNotificationsConcern
+
+  class << self
+    # This is to allow client code to call create notifications without explicitly creating
+    # notifiers.
+    #
+    # @api private
+    def method_missing(symbol, *args, &block)
+      new.send(symbol, *args, &block)
+    end
+  end
+
+  protected
+
+  # Create an ActivityWrapper based on options
+  #
+  # @param [Hash] options The options used to create an activity
+  # @option options [User] :actor The actor who trigger off the activity
+  # @option options :object The object which the activity is about
+  # @option options [Symbol] :event The event name of activity
+  def create_activity(options)
+    ActivityWrapper.new(self, Activity.new(options.merge(notifier_type: self.class.name)))
+  end
+
+  private
+
+  def initialize #:nodoc:
+    super
+    @pending_emails = []
+  end
+
+  # Generate emails according to input recipient and notification
+  #
+  # @param [Object] recipient The recipient of the notification
+  # @param [Course::Notification] notification The target notification
+  def notify(recipient, notification)
+    return unless notification.email?
+    case recipient
+    when Course
+      email_course(notification)
+    when User
+      email_user(notification)
```

Comment:
I have added this, but I think this is useless. Because `notify` will only be called if it can generate a notification successfully. If the recipient is invalid, it will raise an error when creating the notification.